File: kotlin/ex2_3.kt
open class Person
class Student : Person()

fun startMentoring(select: (Student) -> Person): Unit {
    val st: Student = Student()
    val mentor: Person = select(st)
}

fun selectStudentMentor(st: Student): Student = Student()

fun sendEmails(needEmail: (Student) -> Boolean): Unit {
    val st: Student = Student()
    if (needEmail(st)) {}
}

fun isHacked(pr: Person): Boolean = true

fun main() {
    startMentoring(::selectStudentMentor)
    sendEmails(::isHacked)
}

File: scala/ex2_3.scala
class Person
class Student extends Person

def startMentoring(select: Student => Person): Unit =
  val st: Student = Student()
  val mentor: Person = select(st)

def selectStudentMentor(st: Student): Student = Student()

def sendEmails(needEmail: Student => Boolean): Unit =
  val st: Student = Student()
  if needEmail(st) then ()

def isHacked(pr: Person): Boolean = true

@main def main =
  startMentoring(selectStudentMentor)
  sendEmails(isHacked)
